### PR TITLE
add vm test

### PR DESCRIPTION
**Pull request by Artturin**

### CODE REVIEW

File: tests/firefox.nix [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+      ];
+  });
+
+in
+{
+  name = firefoxPackage'.unwrapped.pname;
+  nodes.machine =
+    { pkgs, modulesPath, ... }:
+
+    { imports = [ (modulesPath + "/../tests/common/x11.nix") ];
```
Comment (Artturin) [inline tests/firefox.nix:15]: `${pkgs}/nixos/tests/common/x11.nix` gives infinite recursion

**Comment by colemickens** [comment]
I'm a bit ignorant of some of the finer nix/nixos testing things. Is there a preferred way to invoke this from `nix flake ... `?

**Comment by Artturin** [comment]
Changed the attr from tests to checks which are run by `nix flake check`

**Comment by colemickens** [comment]
Hm, not to rush you since it's still a draft, but I seem to be stuck in the sleep 2 loop:

I pulled this and stuck it on top of what should have been a working commit:

```
commit e3dd5f7a722b0a08f6c97ca51ac21bc9db6b9788
Author: Cole Botkens <cole.mickens+colebot@gmail.com>
Date:   Mon Sep 18 12:20:36 2023 +0000

    firefox-nightly-bin: 2023-09-17-21-19-57 -> 2023-09-18-09-41-02
```

```
[snip]
vm-test-run-firefox-release-bin-unwrapped> (finished: waiting for a window to appear, in 4.24 seconds)
vm-test-run-firefox-release-bin-unwrapped> machine: must succeed: sleep 40
vm-test-run-firefox-release-bin-unwrapped> machine: must succeed: xwininfo -root -tree | sed 's/.*0x[0-9a-f]* \"\([^\"]*\)\".*/\1/; t; d'
vm-test-run-firefox-release-bin-unwrapped> (finished: must succeed: xwininfo -root -tree | sed 's/.*0x[0-9a-f]* \"\([^\"]*\)\".*/\1/; t; d', in 0.08 seconds)
vm-test-run-firefox-release-bin-unwrapped> (finished: waiting for a window to appear, in 4.24 seconds)
vm-test-run-firefox-release-bin-unwrapped> machine: must succeed: sleep 40
vm-test-run-firefox-release-bin-unwrapped> machine: must succeed: xwininfo -root -tree | sed 's/.*0x[0-9a-f]* \"\([^\"]*\)\".*/\1/; t; d'
vm-test-run-firefox-release-bin-unwrapped> (finished: must succeed: xwininfo -root -tree | sed 's/.*0x[0-9a-f]* \"\([^\"]*\)\".*/\1/; t; d', in 0.06 seconds)
vm-test-run-firefox-release-bin-unwrapped> (finished: waiting for a window to appear, in 4.23 seconds)
vm-test-run-firefox-release-bin-unwrapped> machine: must succeed: sleep 40
vm-test-run-firefox-release-bin-unwrapped> machine # [   37.918979] systemd[1]: systemd-timedated.service: Deactivated successfully.
vm-test-run-firefox-release-bin-unwrapped> machine # [   37.895675] systemd[1]: systemd-timedated.service: Deactivated successfully.
vm-test-run-firefox-release-bin-unwrapped> (finished: must succeed: sleep 40, in 40.03 seconds)
vm-test-run-firefox-release-bin-unwrapped> (finished: subtest: Wait until Firefox has finished loading the Valgrind docs page, in 44.23 seconds)
vm-test-run-firefox-release-bin-unwrapped> subtest: Check whether Firefox can play sound
vm-test-run-firefox-release-bin-unwrapped> machine # [   51.301380] systemd[1]: Started Record NixOS test audio to /tmp/record.wav.
vm-test-run-firefox-release-bin-unwrapped> machine: must succeed: firefox file:///nix/store/wyg7cyr6d0p5gxdhp4w2z3if14qgfzjw-sound-theme-freedesktop-0.8/share/sounds/freedesktop/stereo/phone-incoming-call.oga >&2 &
vm-test-run-firefox-release-bin-unwrapped> (finished: must succeed: firefox file:///nix/store/wyg7cyr6d0p5gxdhp4w2z3if14qgfzjw-sound-theme-freedesktop-0.8/share/sounds/freedesktop/stereo/phone-incoming-call.oga >&2 &, in 0.01 seconds)
vm-test-run-firefox-release-bin-unwrapped> machine: waiting for file '/tmp/record.wav'
vm-test-run-firefox-release-bin-unwrapped> machine # [   51.316858] audio-recorder-start[1300]: Recording WAVE '/tmp/record.wav' : Signed 16 bit Little Endian, Rate 48000 Hz, Mono
vm-test-run-firefox-release-bin-unwrapped> (finished: must succeed: sleep 40, in 40.03 seconds)
vm-test-run-firefox-release-bin-unwrapped> (finished: subtest: Wait until Firefox has finished loading the Valgrind docs page, in 44.28 seconds)
vm-test-run-firefox-release-bin-unwrapped> subtest: Check whether Firefox can play sound
vm-test-run-firefox-release-bin-unwrapped> (finished: waiting for file '/tmp/record.wav', in 0.02 seconds)
vm-test-run-firefox-release-bin-unwrapped> machine # [   51.367922] systemd[1]: Started Record NixOS test audio to /tmp/record.wav.
vm-test-run-firefox-release-bin-unwrapped> machine: must succeed: firefox file:///nix/store/wyg7cyr6d0p5gxdhp4w2z3if14qgfzjw-sound-theme-freedesktop-0.8/share/sounds/freedesktop/stereo/phone-incoming-call.oga >&2 &
vm-test-run-firefox-release-bin-unwrapped> machine: must succeed: stat -c '%s' /tmp/last
vm-test-run-firefox-release-bin-unwrapped> (finished: must succeed: firefox file:///nix/store/wyg7cyr6d0p5gxdhp4w2z3if14qgfzjw-sound-theme-freedesktop-0.8/share/sounds/freedesktop/stereo/phone-incoming-call.oga >&2 &, in 0.01 seconds)
vm-test-run-firefox-release-bin-unwrapped> machine: waiting for file '/tmp/record.wav'
vm-test-run-firefox-release-bin-unwrapped> machine # [   51.384600] audio-recorder-start[1278]: Recording WAVE '/tmp/record.wav' : Signed 16 bit Little Endian, Rate 48000 Hz, Mono
vm-test-run-firefox-release-bin-unwrapped> (finished: must succeed: stat -c '%s' /tmp/last, in 0.02 seconds)
vm-test-run-firefox-release-bin-unwrapped> (finished: waiting for file '/tmp/record.wav', in 0.02 seconds)
vm-test-run-firefox-release-bin-unwrapped> machine: must succeed: sleep 2
vm-test-run-firefox-release-bin-unwrapped> (finished: must succeed: sleep 40, in 40.03 seconds)
vm-test-run-firefox-release-bin-unwrapped> (finished: subtest: Wait until Firefox has finished loading the Valgrind docs page, in 44.28 seconds)
vm-test-run-firefox-release-bin-unwrapped> subtest: Check whether Firefox can play sound
vm-test-run-firefox-release-bin-unwrapped> machine: must succeed: stat -c '%s' /tmp/last
vm-test-run-firefox-release-bin-unwrapped> machine # [   51.466549] systemd[1]: Started Record NixOS test audio to /tmp/record.wav.
vm-test-run-firefox-release-bin-unwrapped> machine: must succeed: firefox file:///nix/store/wyg7cyr6d0p5gxdhp4w2z3if14qgfzjw-sound-theme-freedesktop-0.8/share/sounds/freedesktop/stereo/phone-incoming-call.oga >&2 &
vm-test-run-firefox-release-bin-unwrapped> (finished: must succeed: stat -c '%s' /tmp/last, in 0.02 seconds)
vm-test-run-firefox-release-bin-unwrapped> (finished: must succeed: firefox file:///nix/store/wyg7cyr6d0p5gxdhp4w2z3if14qgfzjw-sound-theme-freedesktop-0.8/share/sounds/freedesktop/stereo/phone-incoming-call.oga >&2 &, in 0.01 seconds)
vm-test-run-firefox-release-bin-unwrapped> machine: waiting for file '/tmp/record.wav'
vm-test-run-firefox-release-bin-unwrapped> machine # [   51.484290] audio-recorder-start[1327]: Recording WAVE '/tmp/record.wav' : Signed 16 bit Little Endian, Rate 48000 Hz, Mono
vm-test-run-firefox-release-bin-unwrapped> (finished: waiting for file '/tmp/record.wav', in 0.02 seconds)
vm-test-run-firefox-release-bin-unwrapped> machine: must succeed: sleep 2
vm-test-run-firefox-release-bin-unwrapped> machine: must succeed: stat -c '%s' /tmp/last
vm-test-run-firefox-release-bin-unwrapped> (finished: must succeed: stat -c '%s' /tmp/last, in 0.02 seconds)
vm-test-run-firefox-release-bin-unwrapped> machine: must succeed: sleep 2
vm-test-run-firefox-release-bin-unwrapped> (finished: must succeed: sleep 40, in 40.04 seconds)
vm-test-run-firefox-release-bin-unwrapped> (finished: subtest: Wait until Firefox has finished loading the Valgrind docs page, in 44.28 seconds)
vm-test-run-firefox-release-bin-unwrapped> subtest: Check whether Firefox can play sound
vm-test-run-firefox-release-bin-unwrapped> machine # [   52.246370] systemd[1]: Started Record NixOS test audio to /tmp/record.wav.
vm-test-run-firefox-release-bin-unwrapped> machine: must succeed: firefox file:///nix/store/wyg7cyr6d0p5gxdhp4w2z3if14qgfzjw-sound-theme-freedesktop-0.8/share/sounds/freedesktop/stereo/phone-incoming-call.oga >&2 &
vm-test-run-firefox-release-bin-unwrapped> (finished: must succeed: firefox file:///nix/store/wyg7cyr6d0p5gxdhp4w2z3if14qgfzjw-sound-theme-freedesktop-0.8/share/sounds/freedesktop/stereo/phone-incoming-call.oga >&2 &, in 0.01 seconds)
vm-test-run-firefox-release-bin-unwrapped> machine: waiting for file '/tmp/record.wav'
vm-test-run-firefox-release-bin-unwrapped> machine # [   52.262889] audio-recorder-start[1321]: Recording WAVE '/tmp/record.wav' : Signed 16 bit Little Endian, Rate 48000 Hz, Mono
vm-test-run-firefox-release-bin-unwrapped> (finished: waiting for file '/tmp/record.wav', in 0.02 seconds)
vm-test-run-firefox-release-bin-unwrapped> machine: must succeed: stat -c '%s' /tmp/last
vm-test-run-firefox-release-bin-unwrapped> (finished: must succeed: stat -c '%s' /tmp/last, in 0.02 seconds)
vm-test-run-firefox-release-bin-unwrapped> machine: must succeed: sleep 2
vm-test-run-firefox-release-bin-unwrapped> (finished: must succeed: sleep 2, in 2.02 seconds)
```

**Comment by Artturin** [comment]
Rebased on to that commit

**Comment by Artturin** [comment]
`      packages = { x86_64-linux = { firefox = pkgs_.nixpkgs.x86_64-linux.firefox; }; };`

`nix build ".#checks.x86_64-linux.firefox"`

works so should be something in the firefox versions herer

**Comment by Artturin** [comment]
Debugging with `checks.x86_64-linux.firefox-nightly-bin.driverInteractive`


image

**Comment by colemickens** [comment]
Huh.

**Comment by Artturin** [comment]
Ran the nixpkgs Firefox test interactively and I can see the media bar move and can pause and resume but on `firefox-nightly` I can't see it move, and it's at 1, I can move it but i can't unpause it.

**Comment by Artturin** [comment]
Could be related 
image

**Comment by Artturin** [comment]
`phone-incoming-call.oga` works outside so maybe missing libs?

**Comment by Artturin** [comment]
No diffs other than the hash in firefox-bin `nix-diff $(nix eval --raw ".#firefox-bin.drvPath") $(nix eval --raw "nixpkgs/ace5093e36ab1e95cb9463863491bee90d5a4183#firefox-bin.drvPath")`

**Comment by Artturin** [comment]
`      packages = { x86_64-linux = { firefox = pkgs_.nixpkgs.x86_64-linux.firefox-beta-bin; }; };`

fails too

**Comment by Artturin** [comment]
`nix-diff $(nix eval --raw ".#firefox-beta-bin.drvPath") $(nix eval --raw "nixpkgs/ace5093e36ab1e95cb9463863491bee90d5a4183#firefox-beta.drvPath")`

image

**Comment by Artturin** [comment]
```diff
diff --git a/pkgs/applications/networking/browsers/firefox/wrapper.nix b/pkgs/applications/networking/browsers/firefox/wrapper.nix
index e909b15f77ae..cd764c0150eb 100644
--- a/pkgs/applications/networking/browsers/firefox/wrapper.nix
+++ b/pkgs/applications/networking/browsers/firefox/wrapper.nix
@@ -54,10 +54,10 @@ let
     let
       ffmpegSupport = browser.ffmpegSupport or false;
       gssSupport = browser.gssSupport or false;
-      alsaSupport = browser.alsaSupport or false;
-      pipewireSupport = browser.pipewireSupport or false;
-      sndioSupport = browser.sndioSupport or false;
-      jackSupport = browser.jackSupport or false;
+      alsaSupport = browser.alsaSupport or true;
+      pipewireSupport = browser.pipewireSupport or true;
+      sndioSupport = browser.sndioSupport or true;
+      jackSupport = browser.jackSupport or true;
       # PCSC-Lite daemon (services.pcscd) also must be enabled for firefox to access smartcards
       smartcardSupport = cfg.smartcardSupport or false;
 ```
 
 The diffs are now almost the same but it still doesn't work

**Comment by Artturin** [comment]
Does the firefox distributed by mozilla only support pulseaudio?

**Comment by Artturin** [comment]
Audio tests would be nice but they're not necessary for just making sure the browser can open

**Comment by colemickens** [comment]
@Artturin I use Pipewire and firefox-nightly has been working for me. I agree that punting on the audio issue is fine for now, though it's going to sit in the back of my head for a while.

I fear I've lost push access after transferring this (I've asked Jonas to fix my permissions in the matrix room), but I'll try to revert some commits and merge this ASAP. (And add you as a maintainer.)

**Comment by colemickens** [comment]
Testing:

1. Looks like if I revert to LKG, this test passes, but...
2. Looks like if I drop the reverts and stick this on master, it still infinite loops.

I waited a minute or two and it's "stuck" waiting for `xwininfo`:

```
vm-test-run-firefox-release-bin-unwrapped> machine: must succeed: xwininfo -root -tree | sed 's/.*0x[0-9a-f]* \"\([^\"]*\)\".*/\1/; t; d'
vm-test-run-firefox-release-bin-unwrapped> (finished: must succeed: xwininfo -root -tree | sed 's/.*0x[0-9a-f]* \"\([^\"]*\)\".*/\1/; t; d', in 0.01 seconds)
vm-test-run-firefox-release-bin-unwrapped> machine: must succeed: xwininfo -root -tree | sed 's/.*0x[0-9a-f]* \"\([^\"]*\)\".*/\1/; t; d'
vm-test-run-firefox-release-bin-unwrapped> (finished: must succeed: xwininfo -root -tree | sed 's/.*0x[0-9a-f]* \"\([^\"]*\)\".*/\1/; t; d', in 0.01 seconds)
vm-test-run-firefox-release-bin-unwrapped> machine: must succeed: xwininfo -root -tree | sed 's/.*0x[0-9a-f]* \"\([^\"]*\)\".*/\1/; t; d'
vm-test-run-firefox-release-bin-unwrapped> (finished: must succeed: xwininfo -root -tree | sed 's/.*0x[0-9a-f]* \"\([^\"]*\)\".*/\1/; t; d', in 0.01 seconds)
```

Can you have it timeout/fail after an appropriate amount of time?

Does `wait_for_window` take a timeout that will fail the test?

**Comment by colemickens** [comment]
I'm going to merge as-is so that we at least keep things in working order. Opened #16 to track short-circuiting the test on failure mode.

**Comment by Artturin** [comment]
> @Artturin I use Pipewire and firefox-nightly has been working for me. I agree that punting on the audio issue is fine for now, though it's going to sit in the back of my head for a while.

yeah but i think it may be working for us because of some impurity

**Comment by colemickens** [comment]
Through the emulator though?

**Pull request merged**